Add support for RISC-V standard opcodes

arichardson opened the pull request:
This makes the new instruction encodings available for ISAv9.
I plan to rebase the CHERI-Alliance qemu on top of this in a way that builds two separate simulators.

CC: @martin-kaiser @buxtonpaul

LaneWolf committed:
qemu-img: remove unused variable

Variable block_count used in img_dd function is only incremented but never read.
This causes 'Unused but set variable' warning on Clang 15.0.1 compiler.

Removing the variable to prevent the warning.

Signed-off-by: Miroslav Rezanina <mrezanin@redhat.com>
Reviewed-by: Thomas Huth <thuth@redhat.com>
Message-Id: <e86d5b57f9d13bde995c616a533b876f1fb8a527.1668009030.git.mrezanin@redhat.com>
Signed-off-by: Thomas Huth <thuth@redhat.com>

(cherry picked from commit 0f48c47c679bc29fceb3a67283ee3b78175524bf)

martin-kaiser committed:
riscv: cheri: remove the UTCC SCR register

This register was added as part of the N extension that has been
deprecated, so there is no need to extend it for CHERI.

martin-kaiser committed:
riscv: cheri: remove the UTDC SCR register

This register was added as part of the N extension that has been
deprecated, so there is no need to extend it for CHERI.

martin-kaiser committed:
riscv: cheri: remove the UScratchC SCR

This register was added as part of the N extension that has been
deprecated, so there is no need to extend it for CHERI.

martin-kaiser committed:
riscv: cheri: remove the UEPCC SCR

This register was added as part of the N extension that has been
deprecated, so there is no need to extend it for CHERI.

martin-kaiser committed:
CHERI-RISC-V: Add the gctag instruction

The CHERI standard renames the cgettag instruction to gctag and modifies
the encoding. This instruction reads the tag bit and stores it in a register.

Add a new macro for the translate helpers that allows specifying the
name of the helper so we can reuse the same code.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the gcperm instruction

The CHERI standard renamed the cgetperm instruction to gcperm and modified
the encoding. This instruction reads both AP (architectural permissions)
and SDP (software-defined permissions) and stores them in a register.

Note: This still uses the ISAv9 encoding of permissions, but will
support the new format once we use the cc128r compressed-cap version.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the gchi instruction

The RISC-V standard renamed the cgethigh instruction to gchi and
modified the encoding. This instruction reads the metadata of a
capability and stores it in a register.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the gcbase instruction

The CHERI RISC-V standard renamed the cgetbase instruction to gcbase and
modified the encoding. This instruction reads the base address of a
capability and stores it in a register.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the gclen instruction

The RISC-V standard renamed the CGetLen instruction to gclen and
modified the encoding. This instruction reads the length of a capability
and stores it in a register.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the caddi instruction

The RISC-V standard renamed the CIncOffsetImm instruction to caddi and
modified the encoding. This instruction adds a constant offset to the
address of a capability.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add support for cadd and cmv

The RISC-V standard renamed the CIncOffset instruction to cadd and
modified the encoding. Additionally, the encoding of `cadd cd, cs1, x0`
is used as cmv, so is a move operation where we must not check if cs1 is
tagged or sealed.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the scaddr instruction

The RISC-V standard renamed the csetaddr instruction to scaddr and
modified the encoding. This instruction sets the address field of a
capability.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the schi instruction

The RISC-V standard renamed the csethigh instruction to schi and
modified the encoding. This instruction sets the metadata of a capability.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the sceq instruction

The RISC-V standard renamed the csetequalexact instruction to sceq and
modified the encoding. This instruction compares two capabilities.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: Add #defines for the new CSR numbers

In the RISC-V standard the capability CSRs are accessed using CSRRW
instead of using a new instruction. Add the required defines for this.

buxtonpaul committed:
CHERI-RISC-V: {vs,s,m}scratchc extends {vs,s,m}scratch

This was changed for ISAv9 and is needed for the RISC-V standard.
Also update VMState structure to ensure the VMState is handled correctly.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: initial infrastructure for capability CSR access

This adds an implementation of read/write functions for the mscratch
mscratchc capability that will eventually be used by CSRRW.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: initial handling of capability CSR reads/writes

The RISC-V standard no longer uses a separate instruction to access
capability registers and instead access them using the CSRR*
instructions. This commit adds the initial infrastructure for
translating the CSR accesses using new capability helpers for CSRs that
are capabilities. Currently only mscratch has a handler registers, but
in the future other registers will also be added.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the sentry instruction

The RISC-V standard renamed the CSealEntry instruction to sentry and
modified the encoding. This instruction makes a sealed copy of a capability.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the cbld instruction

The RISC-V standard renamed the CBuildCap instruction to cbld and
modified the encoding. Additionally, the encoding with auth cap == 0
now refers to the NULL register and invalidates the tag of the input
capability. This adds a check for malformed bounds and a TODO for
malformed permissions, but otherwise the checks are identical to v9.

Also removed an incorrect zeroing of the otype that should not have
been part of the v9 code since it's not present in the sail code.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: Add MTVECC CSR handling

The standard uses MTVECC instead of MTCC.
Add custom CSR read/write implementations for mtvecc.
Add HICAUSE definition used for vector range check
Replace usage of mtcc with mtvecc.

buxtonpaul committed:
CHERI-RISC-V: add STVECC CSR handling

The standard renames STCC to STVECC.
Add the read/write functions for CSRRW.

buxtonpaul committed:
CHERI-RISC-V: add MEPCC CSR handling

Reroute access to MEPCC through the CSR register access methods.

buxtonpaul committed:
CHERI-RISC-V: add SEPCC CSR handlers

Add the handling of the SEPCC capability to the CSR handler function.

buxtonpaul committed:
CHERI-RISC-V: add SSCRATCHC CSR capability handling

Adds the handling of SSCARTCHC to the CSR functions.

buxtonpaul committed:
target/riscv: add reset and disas for mseccfg CSR

Add the mseccfg csr register reset to zero (needed for CHERI).
Does not implement any other functionality.

buxtonpaul committed:
riscv: Add menvcfg csr register.

The menvcfg register is required for CHERI but as part of the
privileged spec.
We add it here so that we can later implement the CRE access control in
conjunction with the mseccfg and senvcfg registers.

buxtonpaul committed:
riscv: Add senvcfg csr register.

The senvcfg register is required for CHERI but as part of the
privileged spec.
We add it here so that we can later implement the CRE access control in
conjunction with the mseccfg and menvcfg registers.

buxtonpaul committed:
CHERI-RISC-V: add the scbndsr instruction

The RISC-V standard renames csetbounds to scbndsr and uses a new encoding.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the acperm instruction

The RISC-V standard renamed the CAndPerm instruction to ACPERM and
modified the encoding. The permissions bit mask is different for
ACPERM/GCPERM compared to the ISAv9 versions but that is handled by
the cheri-compressed-cap library.
This is still missing the legalization that is performed for RV32 since
not all encodings are valid.

Co-authored-by: Paul Buxton <paul.buxton@codasip.com>
Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: Use caddi name for compressed capability adds

buxtonpaul committed:
CHERI-RISC-V: add the scbnds instruction

The RISC-V standard renames csetboundsexact and uses a new encoding.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: add handler for the DDC CSR

buxtonpaul committed:
[UPSTREAM] riscv: Update immediate encodings for shift instruction

The slli, srli and srai instructions are defined in the spec to have the
top 6 bits coded to specific values, but implementation in qemu used
just the top 5 bits with the other bits being treated as top of the
immediate value. This was benign until other instructions overlap with
the opcode space in those bits.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: add the scbndsi instruction

The RISC-V standard renames csetboundsimm to scbndsi and uses a new encoding.
Instead of a 12-bit immediate we now use a scaled 5-bit one.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: add the new cram instruction encoding

The RISC-V standard uses a different encoding for CRAM.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>
Co-authored-by: Martin Kaiser <martin.kaiser@codasip.com>

martin-kaiser committed:
cheri: riscv: function to determine pointer mode

This patch adds a function to check if we're running in integer pointer
or capability pointer mode. It read the CRE configuration and the
privilege level.

martin-kaiser committed:
riscv: define the CRE bits

Add defines for cheri's CRE bits. They are used for setting integer or
capability pointer mode.

martin-kaiser committed:
CHERI-RISC-V: add the scmode instruction

The RISC-V standard renamed the CSetFlags instruction to SCMODE and
modified the encoding. Additionally, the value of the mode bit was
inverted and we only update the mode if the execute permission is set.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: check for CRE status on CSR register access.

The newly added capability length registers should only be accessed if
the CRE configuration permits (even when not in capmode).
So we add a flag to the CSR table to indicate this and then use it to
decide when CRE checking is required.

Commit:
CHERI-RISC-V: Fix handling of c.mv instruction

In capability mode the compressed c.mv instruction should be
decoded as cmv cd, cs1. It is supposed to move a capability and not just
a register value.

For now keep the old behaviour for legacy xcheri.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
CHERI-RISC-V: add the scss instruction

The RISC-V standard renamed the CTestSubset instruction to scss and
modified the encoding. Additionally, it no longer allows DDC as an
operand. Since the semantics are slightly different, this is implemented
in a separate helper.

Co-authored-by: Paul Buxton <paul.buxton@codasip.com>
Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: replace a number of CSR write functions with a common one

Reduce code duplication by moving to a common CSR write function where
possible.

buxtonpaul committed:
cheri-common: Add a cap_scaddr function.

Add a helper function which implements the scaddr semantics required
when writing csr registers.

buxtonpaul committed:
CHERI-RISC-V: Add invalid address handling functions.

Add the functions to be used for invalid address handling when writing
CSR registers. The implementation of invalid address handling is
implementation specific. The Implementation here is that found in the
Codasip implementation of CHERI.
We also borrow the SignExtend64 function from Capstone.

buxtonpaul committed:
CHERI-RISC-V: Update handling of vector registers with CSR writes.

buxtonpaul committed:
disas/riscv: add c.lc compressed capmode instruction

buxtonpaul committed:
disas/riscv: add the compressed c.sc capability store instruction

buxtonpaul committed:
disas/riscv: add the c.lcsp compressed capability instruction

buxtonpaul committed:
disas/riscv: add the c.scsp compressed capability instruction

buxtonpaul committed:
CHERI-RISC-V: add CSR registers for Zstid extension

Zstid introduces capability length csr registers to be used for thread ids.
For now the registers are only expected to be used in CHERI emulators,
we do not add support to non-cheri builds for this.

buxtonpaul committed:
disas/riscv: add the rv32 compressed mode variants for c.lc, c.lcsp, c.sc, c.scsp

The opcodes for these overlap with c.fl / c.fs equivalents and are
disambiguated through the int/capmode.

buxtonpaul committed:
CHERI-RISC-V: Update csr_needs_asr for new standard spec

The ASR permissions in the V9 spec operated on a whitelist.
The RISC-V CHERI spec simplifies this by using the privilege mode of the
register to determine ASR requirements.
The TID registers introduced in Zstid are handled slightly differently
as they can be read without ASR, but always require ASR for writes.

In theory, we should guard this by a RISC-V standard #ifdef, but since
this is just a relaxation of the existing restrictions it will not break
any ISAv9 code.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: correctly use read/write flags for CSR permission check

The capability csr access functions were always assuming a write occurred for
the purpose of checking permissions.
Replace this behaviour with check based on source register/immediate.

buxtonpaul committed:
CHERI-RISC-V: Check for asr access using csr_needs_asr

Generate a CHERI exception if the csr access requires asr permissions and it is not present.

martin-kaiser committed:
riscv: cheri: clean up check_csr_cap_permissions

Clean up the check_csr_cap_permissions function.

* Make the function static, it's used only in target/riscv/op_helper_cheri.c.
* Update the comment about return values, the function returns negative
  values other than -1. Add / fix some other comments while at it.
* The int write_mask parameter is used as a boolean value. Change it to
  bool and rename it to write_access. Update the callers accordingly.
  The same goes for the read_only variable.
* Check up front that the csr extension is enabled. If not, there's no
  point in doing any further checks.

martin-kaiser committed:
cheri: riscv: use RISCVException in CHERI CSR helpers

Convert the helper functions for capability-sized CSRs to RISCVException
and the new approach for exceptions.

The "sub-helpers" such as check_csr_cap_permissions do now return
RISCV_EXCP_NONE or a positive exception value.

martin-kaiser committed:
riscv: move cheri trans stubs behind decode-insnXX.c.inc

For non-cheri builds, we still have to add trans_xxx stubs for some cheri
instructions. Move those stubs to some place after decode-insn32.c.inc
and decode-insn16.c.inc are included. The *.c.inc files have typedefs for
the function arguments, we can reuse them for the stubs.

martin-kaiser committed:
riscv: use a macro to generate the translation stubs

Use a macro to generate the stubs for cheri translation functions in
non-cheri builds.

buxtonpaul committed:
CHERI-RISC-V: Replace use of CSR bool fields with a bitfield.

Replace the individual flags in the riscv_csr_cap_ops struct with a
single flags field and #definines indicating the flag.
This is easier to read in the definition of the ops table and uses less space.

martin-kaiser committed:
CHERI-RISC-V: add the gcmode instruction

The RISC-V standard renamed the CGetFlags instruction to GCMODE and
modified the encoding. Additionally, the value of the mode bit was
inverted and we only return the mode if the execute permission is set.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
riscv: csr: make csr_cap_ops static

Make the csr_cap_ops array static. It is used only in csr.c.

martin-kaiser committed:
CHERI-RISC-V: add the gctype instruction

The RISC-V standard renamed the CGetType instruction to GCTYPE and
modified the encoding. Additionally, the value of type bits has changed:
0 now means unsealed and 1 means sentry. This is handled by the
cheri-compressed-cap library, so we can reuse the existing helper.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser committed:
riscv: csr: simplify get_csr_cap_info

Refactor get_csr_cap_info to iterate over the table. This makes it easier
to remove a csr from the table or to add a csr somewhere in the middle.

martin-kaiser committed:
riscv: csr: use consistent types for csr numbers

Use consistent types for csr numbers in get_csr_cap_info and csr_cap_ops.

At the moment, callers of get_csr_cap_info pass the csr number as uint32_t.
get_csr_cap_info's csr number parameter is a signed int.
The csr number csr_cap_ops is an unsigned int.

Change everything to uint32_t.

buxtonpaul committed:
CHERI-RISC-V: Add integer mode implementation of Zstid CSR registers.

CHERI adds a Zstid specifiction which can exist independently of cheri.
The CHERI versions of the other registers have been added previously,
this adds integer mode implementations of the registers.

buxtonpaul committed:
riscv: Add accessor and feature flag for integer mode zstid registers

Access to the stid registers should be protected by a predicate and
should return illegal instruction if an attempt to access them is made
without the feature.
Add a feature flag for this and enable it in cheri configurations.

buxtonpaul committed:
riscv: Swap stid/vstid when switching hypervisor mode.

Add a stid_hs register and use it to swap the values of stid/vstid when
transitioning between hypervisor modes.

buxtonpaul committed:
CHERI-RISC-V: unify encoding for LC instruction across RV32 and RV64

Instead of a separate encoding for RV32 and RV64 the standard
specification now uses the same encoding for RV32 as the RV64 encoding.
Additionally, the standard no longer uses the same encoding (previously
we used the tentative lq encoding) since that has been repurposed for CMO.

This still keeps the old v9 32-bit encoding around but guarded by a new
pred_xcheri() predicate (to allow using the same input file).

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: add the new standard encoding of SC

The 0.9.0 specification unifies the encodings for the SC instruction so
that both RV32 and RV64 use the same (RV64) encoding.
Keep the old RV32 variant for xcheri for now.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: add standard encoding of lr.c

The updated spec unifies the encodings for the lr.c instruction for
RV32 and RV64. Keep the old Xcheri RV32 version for now.

Note: The instruction doesn't appear to have an associated disassembly
for CHERI yet. Will need to add this later.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
CHERI-RISC-V: add standard encoding of sc.c

The updated spec unifies the encodings for the sc.c instruction for
RV32 and RV64. Keep the old Xcheri RV32 version for now.

Note: The instruction doesn't appear to have an associated disassembly
for CHERI yet. Will need to add this later.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

buxtonpaul committed:
riscv: Add state migration entries for thread id registers.

Add stid specific vmstate structure and add the stid registers into this
structure which uses the stid_needed predicate.
Add this structure as a subsection of the cpu structure.

buxtonpaul committed:
Introduce an CSR_OP_EXTENDED_REG and set it on appropriate registers.

The CSR operations are required to behave differently on newly added
capability registers vs registers that have been extended.
So introduce a flag to differentiate them.

buxtonpaul committed:
Update handling of extended csr registers.

The RiscV specificaiton requires that extended and new cheri CSR registers are handled differently.
So we update and rename clip_if_xlen to write_capmode_reg.
We use the udpated function for both extended and new CSRs for csrrw.
Use the updated function only on extended registers for other csr operations.

martin-kaiser committed:
riscv: cheri: merge two if clauses in csrrs_cap helper

Merge two if (rs1) { ... } clauses in the csrrs_cap helper. This makes
the code a bit easier to understand.

martin-kaiser committed:
riscv: cheri: drop unnecessary return statements

raise_cheri_exception_impl throws an exception and does not return. The
fucntion is marked as QEMU_NORETURN.

Remove the unnecessary return statements after raise_cheri_exception_impl
calls.

martin-kaiser committed:
disas/riscv: distinguish between sc.q and sc.c

The sc.q (for RV128) and sc.c (for CHERI) instructions use the same
encoding.

Update the risc-v disassembly code to support sc.c if the translator set
RISCV_DIS_FLAG_CHERI. Distinguish between integer and capability pointer
mode. Add new display formats for sc.c to match the risc-v cheri
specification.

martin-kaiser committed:
disas/riscv: distinguish between amoswap.q and amoswap.c

The amoswap.q (for RV128) and amoswap.c (for CHERI) instructions use the
same encoding.

Update the risc-v disassembly code to support amoswap.c if the
translator set RISCV_DIS_FLAG_CHERI. Distinguish between integer and
capability pointer mode. Add new display formats for amoswap.c.

martin-kaiser committed:
riscv: cheri: merge two if clauses in csrrc_cap helper

Merge two if (rs1) { ... } clauses in the csrrc_cap helper. This makes
the code a bit easier to understand.

martin-kaiser committed:
disas/riscv: distinguish between lr.q and lr.c

The lr.q (for RV128) and lr.c (for CHERI) instructions use the same encoding.

Update the risc-v disassembly code to support lr.c if the translator set
RISCV_DIS_FLAG_CHERI. Distinguish between integer and capability pointer
mode. Add new display formats for lr.c.

buxtonpaul committed:
CHERI-RISC-V: add disas support for scbndsi

srli and scbndsi differ in bit 26 of the opcode. So we use that to
differentiate them.
Scbndsi is a little unique in that it consists of two registers, an
immediate and a flag which shifts the immediate value if present.
Decoding scbndsi requires adding a new codec to decode the operands.
This includes shifting the immediate if the scaled flag is set.
Whilst the scaled flag is in the same position as the rl field and the
immediate matches the positions of rs2 and shamt5, we implement this
with new names to avoid confusion.

buxtonpaul committed:
CHERI-RISC-V: handle reserved scbndsi encodings

The scale=1, imm<=1 encodings are reserved as those immediate values can
be encoded with scale=0.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

arichardson committed:
CHERI-RISC-V: Use cd/cs1/cs2 in the translation code

This makes the instruction definitions closer to what is written in the
spec and slightly improves type safety. I did not touch the loads/stores
since those would require additional instruction formats.

arichardson committed:
CHERI-RISC-V: Use tcg_constant_* in translate helpers

This avoids the need for all the tcg_temp_free() calls

arichardson committed:
CHERI-RISC-V: check for invalid permission encodings in gcmode/scmode

We have to return 0/not update the mode in this case.

arichardson committed:
CHERI-RISC-V: refactor CSR reads to reduce duplication

Reads are mostly routed through to a common function.

buxtonpaul commented:
Can we agree that we avoid rebasing the Cheri-Alliance repo where possible...

Given that decodetree doesn't appear to have a conditional compilation how do you plan on handling the opcodes which appear in the standard riscv insn16 and insn32 .decode files?
I am wondering if perhaps the cleanest way would be to have both the V9 and RiscV Spec encodings present in the same file and use a predicate  that resolves to True/False based on the build options? I don't think there is anything that prevents multiple opcodes decoding to the same trans_xx function..

Worth keeping in mind we will almost certainly have to update these again once RiscV officiales allocates an opcode space for Cheri.

buxtonpaul committed:
CHERI-RISC-V: Rework Capability register CSR write handling.

Implement a new write mechanic capable of handling both capability and
xlen writes to a csr register.
Introduce flags for the registers indicating what actions should be
taken during the write regarding invalid_address handling.
Reduce number of 'special' write functions to reduce code duplication.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

arichardson committed:
CHERI-RISC-V: Implement cmv in tcg without helper calls

Co-authored-by: Paul Buxton <paul.buxton@codasip.com>
Co-authored-by: Martin Kaiser <martin.kaiser@codasip.com>

arichardson committed:
Add new qemu-system-riscv{32,64}std binaries

The plan is to support the ISAv9 format and the new risc-v format from one
source tree (at least temporarily), but using different binaries.

arichardson committed:
Add support for using the cc64r/cc128r capability format

This disables the instructions that rely on otype-based sealing and in
the future will also remove all ISAv9 encodings of instructions.

arichardson committed:
CHERI-RISC-V: split the supported instructions for v9 and std

Loads and stores are not supported by the standard binary yet.

arichardson committed:
CHERI-RISC-V: poison the new identifiers for target-specific code

arichardson commented:
> Can we agree that we avoid rebasing the Cheri-Alliance repo where possible...
> 
> Given that decodetree doesn't appear to have a conditional compilation how do you plan on handling the opcodes which appear in the standard riscv insn16 and insn32 .decode files? I am wondering if perhaps the cleanest way would be to have both the V9 and RiscV Spec encodings present in the same file and use a predicate that resolves to True/False based on the build options? I don't think there is anything that prevents multiple opcodes decoding to the same trans_xx function..
> 
> Worth keeping in mind we will almost certainly have to update these again once RiscV officiales allocates an opcode space for Cheri.

I've pushed some new changes to use the new cc128r/cc64r capability encoding and make the instruction translation conditional based on ifdefs. Still missing the new cssrw/sc/amoswap/lr/sc_c.

I've also rebased the codasip branch on top of this, will push the latest version soon.

jrtc27 commented:
> Can we agree that we avoid rebasing the Cheri-Alliance repo where possible...

I mean, the whole point is to get a cleaner history and codebase for the CHERI-Alliance repo, which by definition means rebasing

buxtonpaul committed:
CHERI-RISC-V: add standard encoding of amoswap.c

The updated spec unifies the encodings for the amoswap.c instruction for
RV32 and RV64. Keep the old Xcheri RV32 version for now.

Note: The instruction doesn't appear to have an associated disassembly
for CHERI. Will need to add this later.

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

martin-kaiser commented:
> 
> I've also rebased the codasip branch on top of this, will push the latest version soon.

I've had a play with the codasip-rebased-20250404 branch.

Compilation fails for risc-v non-cheri, arg_xxx structs are missing for some formats.

otype sign extension is missing brackets and returns wrong values.

cmv does no longer fall back to cadd, this triggers an assertion.

One permission check fails for !invalid, it should fail for invalid.

scbndsi seems to be defined multiple times.

I've uploaded some draft commits to address these issues.
https://github.com/martin-kaiser/qemu/commits/review.25-04-07_codasip-rebased-20250404/

This version passes the morello testsuite, our codasip-internal tests and boots a linux kernel if levels are disabled.
Some of those fixes (especially the cmv) need a proper review...

arichardson commented:
> > I've also rebased the codasip branch on top of this, will push the latest version soon.
> 
> I've had a play with the codasip-rebased-20250404 branch.
> 
> Compilation fails for risc-v non-cheri, arg_xxx structs are missing for some formats.
> 
> otype sign extension is missing brackets and returns wrong values.
> 
> cmv does no longer fall back to cadd, this triggers an assertion.
> 
> One permission check fails for !invalid, it should fail for invalid.
> 
> scbndsi seems to be defined multiple times.
> 
> I've uploaded some draft commits to address these issues. https://github.com/martin-kaiser/qemu/commits/review.25-04-07_codasip-rebased-20250404/
> 
> This version passes the morello testsuite, our codasip-internal tests and boots a linux kernel if levels are disabled. Some of those fixes (especially the cmv) need a proper review...

Thank you! Will take a look shortly

arichardson committed:
Move the cchecktype helper to MIPS

This was never added to RISC-V and should not be in the common code.

arichardson committed:
Fix cap_get_otype_signext() for the RISC-V standard

There is no "sign extension" of otypes for the RISC-V standard since the
special otypes are allocated from zero upwards.

Co-authored-by: Martin Kaiser <martin.kaiser@codasip.com>

arichardson committed:
riscv: cheri: Fix names of the hypervisor SCRs

They no longer use a 'bs' prefix but instead use 'vs'

arichardson committed:
CHERI-RISC-V: simplify write_capmode_reg

arichardson committed:
CHERI-RISC-V: share code for the CSRRC[I]/CSRRS[I] helpers

They are all very similar so we can just add a single branch in a new
helper function.

arichardson committed:
CHERI-RISC-V: Use the existing macros for extended CSRs

This makes it more obvious that we are reusing the same CSR number
rather than allocating a new one (as was the case in older drafts of the
CHERI standard).

arichardson committed:
target/riscv: extract predicate check from riscv_csrrw

This will allow reusing the checks for capability CSRs instead of having
to duplicate it.

buxtonpaul committed:
CHERI-RISC-V: Add CSR number defines for the *TID registers

Co-authored-by: Alexander Richardson <alexrichardson@google.com>

arichardson committed:
CHERI-RISC-V: simplify check_csr_cap_permissions()

Most of the checks are shared with the newly factored out
riscv_csr_accessible(), so we can just call that for most of them.

arichardson committed:
CHERI-RISC-V: raise exceptions in check_csr_cap_permissions()

This greatly simplifies the callers and also avoids the duplicate check
for ASR that was being performed by the helpers as well as
riscv_csr_accessible().

arichardson committed:
CHERI-RISC-V: further simplify CSR helpers

Refactor write_capmode_reg() to writeback_csrrw() and handle the
CSR_OP_EXTENDED_REG flag in this function instead of the callers.

arichardson committed:
CHERI-RISC-V: add CSRRW handlers for the vs* CSRs

We had vstidc, but none of the others.

arichardson committed:
CHERI-RISC-V: Handle capability CSRs in riscv_log_instr_csr_changed()

This allows removing the LOG_SPECIAL_REG() helper macro.

arichardson committed:
CHERI-RISC-V: Use the new CSRRW infrastructure for cspecialrw

We need to keep this instruction around to keep CI with the old ISAv9
binaries working. This commit changes all the SCRs over to using the
new infrastructure.
